Improve ExtrusionLine::simplify, eliminating many very-short extrusion segments which led to blemishes in thin-wall models sliced with Arachne

sethml opened the pull request:
This change makes ExtrusionLine::simplify() able to remove more very-short segments, which arachne seems to tend to produce frequently on thin-walled models with curvature. Interestingly, Cura does not seem to have this problem. From what I can see, when Arachne was ported over from Cura, a significant amount of logic was added to polyline simplification, but some of that logic was flawed. In particular, the simplification code in Cura seems to be in [PolygonRef::simplify()](https://github.com/Ultimaker/libArachne/blob/5f80a8c9558bedb60d18711facab282546da70fc/src/utils/polygon.cpp#L320), but during porting logic was added to avoid removing a point prior to a long segment and instead replace the two prior points with the intersection of the prior and next segments, but only if that intersection does not move the points too much. Unfortunately that test compared the intersection to the point at the other end of the long segment, thus the length test would always fail and the point would never be removed.

I made the following changes:

- When considering replacing two points with the intersection of the adjoining segments, test the distance between the intersection and the two points prev and current, rather than prev and next. This change removes most extrusion blemishes around the perimeter.
- For a closed polygon, allow removing the first/last point (which are the same). This can remove a single extrusion blemish per perimeter. In order to do so, we consider removing the last point, and pull the next and next_next points from the second and third points of the newly-filtered path.

While working with this code I noticed a few potential issues/future improvements:

- As points are removed or replaced, their width is not updated. In most cases this probably results in negligible errors. However, I could imagine that if a point is removed between two nearly-collinear long segments and the segments have different widths, significant error could be introduced. Consider either not merging segments with significant width differences, and/or updating widths with a length-weighted average when merging segments.
- Why is arachne generating so many nearly-zero-length segments in the first place? Consider changing it to no longer do so. Given that the nearly-zero-length segments seem to have a slightly different width than neighboring segments, I suspect that whatever heuristics arachne uses to determine width are generating these.
- The logic in ExtrusionLine::simplify is pretty complex and challenging to reason about. Perhaps going back to simpler simplification logic such as Cura uses would produce more consistent results.

I believe this fixes #2840. Here's a photo of the model from that issue with this change:
![PXL_20231205_231056828](https://github.com/SoftFever/OrcaSlicer/assets/673922/2453f32f-1210-4fcd-b66b-7a4a03814aa1)

sethml commented:
Note: I submitted a pull request upstream as well: https://github.com/prusa3d/PrusaSlicer/pull/11811

SoftFever commented on src/libslic3r/Arachne/utils/ExtrusionLine.cpp:
```diff
@@ -134,12 +140,18 @@ void ExtrusionLine::simplify(const int64_t smallest_line_segment_squared, const
                     || Line::distance_to_infinite_squared(intersection_point, previous.p, current.p) > double(allowed_error_distance_squared)
                     || (intersection_point - previous.p).cast<int64_t>().squaredNorm() > smallest_line_segment_squared  // The intersection point is way too far from the 'previous'
-                    || (intersection_point - next.p).cast<int64_t>().squaredNorm() > smallest_line_segment_squared)     // and 'next' points, so it shouldn't replace 'current'
+                    || (intersection_point - current.p).cast<int64_t>().squaredNorm() > smallest_line_segment_squared)  // and 'current' points, so it shouldn't replace 'current'
```
This change makes sense to me.

SoftFever commented:
it looks like the closed loop is still causing troubles:

![image](https://github.com/SoftFever/OrcaSlicer/assets/103989404/9b645e22-5fe0-412e-a0a3-43bb2d191850)

sethml commented:
> it looks like the closed loop is still causing troubles:

Can you provide your .3mf for that?

There are still cases when the intersection distance test on line 139 decides not to discard points and leaves short segments, just many fewer than before. I could investigate more under what conditions that happens and see if we can fix them.

SoftFever commented:
> > it looks like the closed loop is still causing troubles:
> 
> Can you provide your .3mf for that?
> 
> There are still cases when the intersection distance test on line 139 decides not to discard points and leaves short segments, just many fewer than before. I could investigate more under what conditions that happens and see if we can fix them.

I'm using the project you attached.[Sphere-Shell.3mf.zip](https://github.com/SoftFever/OrcaSlicer/files/13432610/Sphere-Shell.3mf.zip)
At layer 60. It looks like a edge case where the first line segment is tiny segment.
![Screenshot 2023-12-09 at 10 18 08 AM](https://github.com/SoftFever/OrcaSlicer/assets/103989404/a2d79280-9bbf-4165-897c-34c7a5db9ab0)

sethml commented:
> > > it looks like the closed loop is still causing troubles:
> > 
> > Can you provide your .3mf for that?
> > There are still cases when the intersection distance test on line 139 decides not to discard points and leaves short segments, just many fewer than before. I could investigate more under what conditions that happens and see if we can fix them.
> 
> I'm using the project you attached.[Sphere-Shell.3mf.zip](https://github.com/SoftFever/OrcaSlicer/files/13432610/Sphere-Shell.3mf.zip)

Oh, oops, I confused myself about which .3mf I uploaded with the bug report!

> At layer 60. It looks like a edge case where the first line segment is tiny segment.

Hmm, I ran this with some debugging output, and I see the tiny segment in the gcode for layer 60, but the perimeter data for the layer doesn't seem to contain it. Here's my debugging data for layer 60:

```
BEFORE ExtrusionLine::simplify on 109 points:
  0: -48.172, 0.000, w 0.369, len 0.000
  1: -48.171, 0.016, w 0.368, len 0.016
  2: -47.442, 8.349, w 0.368, len 8.365
  3: -47.440, 8.365, w 0.369, len 0.016
  4: -47.436, 8.381, w 0.368, len 0.016
  5: -45.271, 16.460, w 0.368, len 8.365
  6: -45.267, 16.476, w 0.369, len 0.016
  7: -45.260, 16.490, w 0.368, len 0.016
  8: -41.725, 24.071, w 0.368, len 8.365
  9: -41.718, 24.086, w 0.369, len 0.016
  10: -41.709, 24.099, w 0.368, len 0.016
  11: -36.911, 30.951, w 0.368, len 8.365
  12: -36.902, 30.964, w 0.369, len 0.016
  13: -36.890, 30.976, w 0.368, len 0.016
  14: -30.976, 36.890, w 0.368, len 8.365
  15: -30.964, 36.902, w 0.369, len 0.016
  16: -30.951, 36.911, w 0.368, len 0.016
  17: -24.099, 41.709, w 0.368, len 8.365
  18: -24.086, 41.718, w 0.369, len 0.016
  19: -24.071, 41.725, w 0.368, len 0.016
  20: -16.490, 45.260, w 0.368, len 8.365
  21: -16.476, 45.267, w 0.369, len 0.016
  22: -16.460, 45.271, w 0.368, len 0.016
  23: -8.381, 47.436, w 0.368, len 8.365
  24: -8.365, 47.440, w 0.369, len 0.016
  25: -8.349, 47.442, w 0.368, len 0.016
  26: -0.016, 48.171, w 0.368, len 8.365
  27: 0.000, 48.172, w 0.369, len 0.016
  28: 0.016, 48.171, w 0.368, len 0.016
  29: 8.349, 47.442, w 0.368, len 8.365
  30: 8.365, 47.440, w 0.369, len 0.016
  31: 8.381, 47.436, w 0.368, len 0.016
  32: 16.460, 45.271, w 0.368, len 8.365
  33: 16.476, 45.267, w 0.369, len 0.016
  34: 16.490, 45.260, w 0.368, len 0.016
  35: 24.071, 41.725, w 0.368, len 8.365
  36: 24.086, 41.718, w 0.369, len 0.016
  37: 24.099, 41.709, w 0.368, len 0.016
  38: 30.951, 36.911, w 0.368, len 8.365
  39: 30.964, 36.902, w 0.369, len 0.016
  40: 30.976, 36.890, w 0.368, len 0.016
  41: 36.890, 30.976, w 0.368, len 8.365
  42: 36.902, 30.964, w 0.369, len 0.016
  43: 36.911, 30.951, w 0.368, len 0.016
  44: 41.709, 24.099, w 0.368, len 8.365
  45: 41.718, 24.086, w 0.369, len 0.016
  46: 41.725, 24.071, w 0.368, len 0.016
  47: 45.260, 16.490, w 0.368, len 8.365
  48: 45.267, 16.476, w 0.369, len 0.016
  49: 45.271, 16.460, w 0.368, len 0.016
  50: 47.436, 8.381, w 0.368, len 8.365
  51: 47.440, 8.365, w 0.369, len 0.016
  52: 47.442, 8.349, w 0.368, len 0.016
  53: 48.171, 0.016, w 0.368, len 8.365
  54: 48.172, 0.000, w 0.369, len 0.016
  55: 48.171, -0.016, w 0.368, len 0.016
  56: 47.442, -8.349, w 0.368, len 8.365
  57: 47.440, -8.365, w 0.369, len 0.016
  58: 47.436, -8.381, w 0.368, len 0.016
  59: 45.271, -16.460, w 0.368, len 8.365
  60: 45.267, -16.476, w 0.369, len 0.016
  61: 45.260, -16.490, w 0.368, len 0.016
  62: 41.725, -24.071, w 0.368, len 8.365
  63: 41.718, -24.086, w 0.369, len 0.016
  64: 41.709, -24.099, w 0.368, len 0.016
  65: 36.911, -30.951, w 0.368, len 8.365
  66: 36.902, -30.964, w 0.369, len 0.016
  67: 36.890, -30.976, w 0.368, len 0.016
  68: 30.976, -36.890, w 0.368, len 8.365
  69: 30.964, -36.902, w 0.369, len 0.016
  70: 30.951, -36.911, w 0.368, len 0.016
  71: 24.099, -41.709, w 0.368, len 8.365
  72: 24.086, -41.718, w 0.369, len 0.016
  73: 24.071, -41.725, w 0.368, len 0.016
  74: 16.490, -45.260, w 0.368, len 8.365
  75: 16.476, -45.267, w 0.369, len 0.016
  76: 16.460, -45.271, w 0.368, len 0.016
  77: 8.381, -47.436, w 0.368, len 8.365
  78: 8.365, -47.440, w 0.369, len 0.016
  79: 8.349, -47.442, w 0.368, len 0.016
  80: 0.016, -48.171, w 0.368, len 8.365
  81: 0.000, -48.172, w 0.369, len 0.016
  82: -0.016, -48.171, w 0.368, len 0.016
  83: -8.349, -47.442, w 0.368, len 8.365
  84: -8.365, -47.440, w 0.369, len 0.016
  85: -8.381, -47.436, w 0.368, len 0.016
  86: -16.460, -45.271, w 0.368, len 8.365
  87: -16.476, -45.267, w 0.369, len 0.016
  88: -16.490, -45.260, w 0.368, len 0.016
  89: -24.071, -41.725, w 0.368, len 8.365
  90: -24.086, -41.718, w 0.369, len 0.016
  91: -24.099, -41.709, w 0.368, len 0.016
  92: -30.951, -36.911, w 0.368, len 8.365
  93: -30.964, -36.902, w 0.369, len 0.016
  94: -30.976, -36.890, w 0.368, len 0.016
  95: -36.890, -30.976, w 0.368, len 8.365
  96: -36.902, -30.964, w 0.369, len 0.016
  97: -36.911, -30.951, w 0.368, len 0.016
  98: -41.709, -24.099, w 0.368, len 8.365
  99: -41.718, -24.086, w 0.369, len 0.016
  100: -41.725, -24.071, w 0.368, len 0.016
  101: -45.260, -16.490, w 0.368, len 8.365
  102: -45.267, -16.476, w 0.369, len 0.016
  103: -45.271, -16.460, w 0.368, len 0.016
  104: -47.436, -8.381, w 0.368, len 8.365
  105: -47.440, -8.365, w 0.369, len 0.016
  106: -47.442, -8.349, w 0.368, len 0.016
  107: -48.171, -0.016, w 0.368, len 8.365
  108: -48.172, 0.000, w 0.369, len 0.016
SIMPLIFYING:
  Replaced 0 with: 1, at -48.172, 0.002
  2 => 1
  Next segment small: 3, next 0.016 < 1.000
  Replaced 1 with: 4, at -47.440, 8.365
  5 => 2
  Next segment small: 6, next 0.016 < 1.000
  Replaced 2 with: 7, at -45.267, 16.476
  8 => 3
  Next segment small: 9, next 0.016 < 1.000
  Replaced 3 with: 10, at -41.718, 24.086
  11 => 4
  Next segment small: 12, next 0.016 < 1.000
  Replaced 4 with: 13, at -36.902, 30.964
  14 => 5
  Next segment small: 15, next 0.016 < 1.000
  Replaced 5 with: 16, at -30.964, 36.902
  17 => 6
  Next segment small: 18, next 0.016 < 1.000
  Replaced 6 with: 19, at -24.086, 41.718
  20 => 7
  Next segment small: 21, next 0.016 < 1.000
  Replaced 7 with: 22, at -16.476, 45.267
  23 => 8
  Next segment small: 24, next 0.016 < 1.000
  Replaced 8 with: 25, at -8.365, 47.440
  26 => 9
  Next segment small: 27, next 0.016 < 1.000
  Replaced 9 with: 28, at 0.000, 48.172
  29 => 10
  Next segment small: 30, next 0.016 < 1.000
  Replaced 10 with: 31, at 8.365, 47.440
  32 => 11
  Next segment small: 33, next 0.016 < 1.000
  Replaced 11 with: 34, at 16.476, 45.267
  35 => 12
  Next segment small: 36, next 0.016 < 1.000
  Replaced 12 with: 37, at 24.086, 41.718
  38 => 13
  Next segment small: 39, next 0.016 < 1.000
  Replaced 13 with: 40, at 30.964, 36.902
  41 => 14
  Next segment small: 42, next 0.016 < 1.000
  Replaced 14 with: 43, at 36.902, 30.964
  44 => 15
  Next segment small: 45, next 0.016 < 1.000
  Replaced 15 with: 46, at 41.718, 24.086
  47 => 16
  Next segment small: 48, next 0.016 < 1.000
  Replaced 16 with: 49, at 45.267, 16.476
  50 => 17
  Next segment small: 51, next 0.016 < 1.000
  Replaced 17 with: 52, at 47.440, 8.365
  53 => 18
  Next segment small: 54, next 0.016 < 1.000
  Replaced 18 with: 55, at 48.172, 0.000
  56 => 19
  Next segment small: 57, next 0.016 < 1.000
  Replaced 19 with: 58, at 47.440, -8.365
  59 => 20
  Next segment small: 60, next 0.016 < 1.000
  Replaced 20 with: 61, at 45.267, -16.476
  62 => 21
  Next segment small: 63, next 0.016 < 1.000
  Replaced 21 with: 64, at 41.718, -24.086
  65 => 22
  Next segment small: 66, next 0.016 < 1.000
  Replaced 22 with: 67, at 36.902, -30.964
  68 => 23
  Next segment small: 69, next 0.016 < 1.000
  Replaced 23 with: 70, at 30.964, -36.902
  71 => 24
  Next segment small: 72, next 0.016 < 1.000
  Replaced 24 with: 73, at 24.086, -41.718
  74 => 25
  Next segment small: 75, next 0.016 < 1.000
  Replaced 25 with: 76, at 16.476, -45.267
  77 => 26
  Next segment small: 78, next 0.016 < 1.000
  Replaced 26 with: 79, at 8.365, -47.440
  80 => 27
  Next segment small: 81, next 0.016 < 1.000
  Replaced 27 with: 82, at 0.000, -48.172
  83 => 28
  Next segment small: 84, next 0.016 < 1.000
  Replaced 28 with: 85, at -8.365, -47.440
  86 => 29
  Next segment small: 87, next 0.016 < 1.000
  Replaced 29 with: 88, at -16.476, -45.267
  89 => 30
  Next segment small: 90, next 0.016 < 1.000
  Replaced 30 with: 91, at -24.086, -41.718
  92 => 31
  Next segment small: 93, next 0.016 < 1.000
  Replaced 31 with: 94, at -30.964, -36.902
  95 => 32
  Next segment small: 96, next 0.016 < 1.000
  Replaced 32 with: 97, at -36.902, -30.964
  98 => 33
  Next segment small: 99, next 0.016 < 1.000
  Replaced 33 with: 100, at -41.718, -24.086
  101 => 34
  Next segment small: 102, next 0.016 < 1.000
  Replaced 34 with: 103, at -45.267, -16.476
  104 => 35
  Next segment small: 105, next 0.016 < 1.000
  Replaced 35 with: 106, at -47.440, -8.365
  107 => 36
  Replaced 36 with: 108, at -48.172, -0.000
AFTER ExtrusionLine::simplify on 37 points:
  0: -48.172, -0.000, w 0.368, len 0.000
  1: -47.440, 8.365, w 0.368, len 8.397
  2: -45.267, 16.476, w 0.368, len 8.397
  3: -41.718, 24.086, w 0.368, len 8.397
  4: -36.902, 30.964, w 0.368, len 8.397
  5: -30.964, 36.902, w 0.368, len 8.397
  6: -24.086, 41.718, w 0.368, len 8.397
  7: -16.476, 45.267, w 0.368, len 8.397
  8: -8.365, 47.440, w 0.368, len 8.397
  9: 0.000, 48.172, w 0.368, len 8.397
  10: 8.365, 47.440, w 0.368, len 8.397
  11: 16.476, 45.267, w 0.368, len 8.397
  12: 24.086, 41.718, w 0.368, len 8.397
  13: 30.964, 36.902, w 0.368, len 8.397
  14: 36.902, 30.964, w 0.368, len 8.397
  15: 41.718, 24.086, w 0.368, len 8.397
  16: 45.267, 16.476, w 0.368, len 8.397
  17: 47.440, 8.365, w 0.368, len 8.397
  18: 48.172, 0.000, w 0.368, len 8.397
  19: 47.440, -8.365, w 0.368, len 8.397
  20: 45.267, -16.476, w 0.368, len 8.397
  21: 41.718, -24.086, w 0.368, len 8.397
  22: 36.902, -30.964, w 0.368, len 8.397
  23: 30.964, -36.902, w 0.368, len 8.397
  24: 24.086, -41.718, w 0.368, len 8.397
  25: 16.476, -45.267, w 0.368, len 8.397
  26: 8.365, -47.440, w 0.368, len 8.397
  27: 0.000, -48.172, w 0.368, len 8.397
  28: -8.365, -47.440, w 0.368, len 8.397
  29: -16.476, -45.267, w 0.368, len 8.397
  30: -24.086, -41.718, w 0.368, len 8.397
  31: -30.964, -36.902, w 0.368, len 8.397
  32: -36.902, -30.964, w 0.368, len 8.397
  33: -41.718, -24.086, w 0.368, len 8.397
  34: -45.267, -16.476, w 0.368, len 8.397
  35: -47.440, -8.365, w 0.368, len 8.397
  36: -48.172, -0.000, w 0.368, len 8.397
```

For each point 'len' is the distance to the previous point. The first point just says 'len 0' since there is no previous point. In any case, I see no tiny segment on this perimeter, so I suspect it comes from some later processing stage. It probably does make the seam slightly worse on that layer.

SoftFever commented:
> > > > it looks like the closed loop is still causing troubles:
> > > 
> > > 
> > > Can you provide your .3mf for that?
> > > There are still cases when the intersection distance test on line 139 decides not to discard points and leaves short segments, just many fewer than before. I could investigate more under what conditions that happens and see if we can fix them.
> > 
> > 
> > I'm using the project you attached.[Sphere-Shell.3mf.zip](https://github.com/SoftFever/OrcaSlicer/files/13432610/Sphere-Shell.3mf.zip)
> 
> Oh, oops, I confused myself about which .3mf I uploaded with the bug report!
> 
> > At layer 60. It looks like a edge case where the first line segment is tiny segment.
> 
> Hmm, I ran this with some debugging output, and I see the tiny segment in the gcode for layer 60, but the perimeter data for the layer doesn't seem to contain it. Here's my debugging data for layer 60:
> 
> ```
> BEFORE ExtrusionLine::simplify on 109 points:
>   0: -48.172, 0.000, w 0.369, len 0.000
>   1: -48.171, 0.016, w 0.368, len 0.016
>   2: -47.442, 8.349, w 0.368, len 8.365
>   3: -47.440, 8.365, w 0.369, len 0.016
>   4: -47.436, 8.381, w 0.368, len 0.016
>   5: -45.271, 16.460, w 0.368, len 8.365
>   6: -45.267, 16.476, w 0.369, len 0.016
>   7: -45.260, 16.490, w 0.368, len 0.016
>   8: -41.725, 24.071, w 0.368, len 8.365
>   9: -41.718, 24.086, w 0.369, len 0.016
>   10: -41.709, 24.099, w 0.368, len 0.016
>   11: -36.911, 30.951, w 0.368, len 8.365
>   12: -36.902, 30.964, w 0.369, len 0.016
>   13: -36.890, 30.976, w 0.368, len 0.016
>   14: -30.976, 36.890, w 0.368, len 8.365
>   15: -30.964, 36.902, w 0.369, len 0.016
>   16: -30.951, 36.911, w 0.368, len 0.016
>   17: -24.099, 41.709, w 0.368, len 8.365
>   18: -24.086, 41.718, w 0.369, len 0.016
>   19: -24.071, 41.725, w 0.368, len 0.016
>   20: -16.490, 45.260, w 0.368, len 8.365
>   21: -16.476, 45.267, w 0.369, len 0.016
>   22: -16.460, 45.271, w 0.368, len 0.016
>   23: -8.381, 47.436, w 0.368, len 8.365
>   24: -8.365, 47.440, w 0.369, len 0.016
>   25: -8.349, 47.442, w 0.368, len 0.016
>   26: -0.016, 48.171, w 0.368, len 8.365
>   27: 0.000, 48.172, w 0.369, len 0.016
>   28: 0.016, 48.171, w 0.368, len 0.016
>   29: 8.349, 47.442, w 0.368, len 8.365
>   30: 8.365, 47.440, w 0.369, len 0.016
>   31: 8.381, 47.436, w 0.368, len 0.016
>   32: 16.460, 45.271, w 0.368, len 8.365
>   33: 16.476, 45.267, w 0.369, len 0.016
>   34: 16.490, 45.260, w 0.368, len 0.016
>   35: 24.071, 41.725, w 0.368, len 8.365
>   36: 24.086, 41.718, w 0.369, len 0.016
>   37: 24.099, 41.709, w 0.368, len 0.016
>   38: 30.951, 36.911, w 0.368, len 8.365
>   39: 30.964, 36.902, w 0.369, len 0.016
>   40: 30.976, 36.890, w 0.368, len 0.016
>   41: 36.890, 30.976, w 0.368, len 8.365
>   42: 36.902, 30.964, w 0.369, len 0.016
>   43: 36.911, 30.951, w 0.368, len 0.016
>   44: 41.709, 24.099, w 0.368, len 8.365
>   45: 41.718, 24.086, w 0.369, len 0.016
>   46: 41.725, 24.071, w 0.368, len 0.016
>   47: 45.260, 16.490, w 0.368, len 8.365
>   48: 45.267, 16.476, w 0.369, len 0.016
>   49: 45.271, 16.460, w 0.368, len 0.016
>   50: 47.436, 8.381, w 0.368, len 8.365
>   51: 47.440, 8.365, w 0.369, len 0.016
>   52: 47.442, 8.349, w 0.368, len 0.016
>   53: 48.171, 0.016, w 0.368, len 8.365
>   54: 48.172, 0.000, w 0.369, len 0.016
>   55: 48.171, -0.016, w 0.368, len 0.016
>   56: 47.442, -8.349, w 0.368, len 8.365
>   57: 47.440, -8.365, w 0.369, len 0.016
>   58: 47.436, -8.381, w 0.368, len 0.016
>   59: 45.271, -16.460, w 0.368, len 8.365
>   60: 45.267, -16.476, w 0.369, len 0.016
>   61: 45.260, -16.490, w 0.368, len 0.016
>   62: 41.725, -24.071, w 0.368, len 8.365
>   63: 41.718, -24.086, w 0.369, len 0.016
>   64: 41.709, -24.099, w 0.368, len 0.016
>   65: 36.911, -30.951, w 0.368, len 8.365
>   66: 36.902, -30.964, w 0.369, len 0.016
>   67: 36.890, -30.976, w 0.368, len 0.016
>   68: 30.976, -36.890, w 0.368, len 8.365
>   69: 30.964, -36.902, w 0.369, len 0.016
>   70: 30.951, -36.911, w 0.368, len 0.016
>   71: 24.099, -41.709, w 0.368, len 8.365
>   72: 24.086, -41.718, w 0.369, len 0.016
>   73: 24.071, -41.725, w 0.368, len 0.016
>   74: 16.490, -45.260, w 0.368, len 8.365
>   75: 16.476, -45.267, w 0.369, len 0.016
>   76: 16.460, -45.271, w 0.368, len 0.016
>   77: 8.381, -47.436, w 0.368, len 8.365
>   78: 8.365, -47.440, w 0.369, len 0.016
>   79: 8.349, -47.442, w 0.368, len 0.016
>   80: 0.016, -48.171, w 0.368, len 8.365
>   81: 0.000, -48.172, w 0.369, len 0.016
>   82: -0.016, -48.171, w 0.368, len 0.016
>   83: -8.349, -47.442, w 0.368, len 8.365
>   84: -8.365, -47.440, w 0.369, len 0.016
>   85: -8.381, -47.436, w 0.368, len 0.016
>   86: -16.460, -45.271, w 0.368, len 8.365
>   87: -16.476, -45.267, w 0.369, len 0.016
>   88: -16.490, -45.260, w 0.368, len 0.016
>   89: -24.071, -41.725, w 0.368, len 8.365
>   90: -24.086, -41.718, w 0.369, len 0.016
>   91: -24.099, -41.709, w 0.368, len 0.016
>   92: -30.951, -36.911, w 0.368, len 8.365
>   93: -30.964, -36.902, w 0.369, len 0.016
>   94: -30.976, -36.890, w 0.368, len 0.016
>   95: -36.890, -30.976, w 0.368, len 8.365
>   96: -36.902, -30.964, w 0.369, len 0.016
>   97: -36.911, -30.951, w 0.368, len 0.016
>   98: -41.709, -24.099, w 0.368, len 8.365
>   99: -41.718, -24.086, w 0.369, len 0.016
>   100: -41.725, -24.071, w 0.368, len 0.016
>   101: -45.260, -16.490, w 0.368, len 8.365
>   102: -45.267, -16.476, w 0.369, len 0.016
>   103: -45.271, -16.460, w 0.368, len 0.016
>   104: -47.436, -8.381, w 0.368, len 8.365
>   105: -47.440, -8.365, w 0.369, len 0.016
>   106: -47.442, -8.349, w 0.368, len 0.016
>   107: -48.171, -0.016, w 0.368, len 8.365
>   108: -48.172, 0.000, w 0.369, len 0.016
> SIMPLIFYING:
>   Replaced 0 with: 1, at -48.172, 0.002
>   2 => 1
>   Next segment small: 3, next 0.016 < 1.000
>   Replaced 1 with: 4, at -47.440, 8.365
>   5 => 2
>   Next segment small: 6, next 0.016 < 1.000
>   Replaced 2 with: 7, at -45.267, 16.476
>   8 => 3
>   Next segment small: 9, next 0.016 < 1.000
>   Replaced 3 with: 10, at -41.718, 24.086
>   11 => 4
>   Next segment small: 12, next 0.016 < 1.000
>   Replaced 4 with: 13, at -36.902, 30.964
>   14 => 5
>   Next segment small: 15, next 0.016 < 1.000
>   Replaced 5 with: 16, at -30.964, 36.902
>   17 => 6
>   Next segment small: 18, next 0.016 < 1.000
>   Replaced 6 with: 19, at -24.086, 41.718
>   20 => 7
>   Next segment small: 21, next 0.016 < 1.000
>   Replaced 7 with: 22, at -16.476, 45.267
>   23 => 8
>   Next segment small: 24, next 0.016 < 1.000
>   Replaced 8 with: 25, at -8.365, 47.440
>   26 => 9
>   Next segment small: 27, next 0.016 < 1.000
>   Replaced 9 with: 28, at 0.000, 48.172
>   29 => 10
>   Next segment small: 30, next 0.016 < 1.000
>   Replaced 10 with: 31, at 8.365, 47.440
>   32 => 11
>   Next segment small: 33, next 0.016 < 1.000
>   Replaced 11 with: 34, at 16.476, 45.267
>   35 => 12
>   Next segment small: 36, next 0.016 < 1.000
>   Replaced 12 with: 37, at 24.086, 41.718
>   38 => 13
>   Next segment small: 39, next 0.016 < 1.000
>   Replaced 13 with: 40, at 30.964, 36.902
>   41 => 14
>   Next segment small: 42, next 0.016 < 1.000
>   Replaced 14 with: 43, at 36.902, 30.964
>   44 => 15
>   Next segment small: 45, next 0.016 < 1.000
>   Replaced 15 with: 46, at 41.718, 24.086
>   47 => 16
>   Next segment small: 48, next 0.016 < 1.000
>   Replaced 16 with: 49, at 45.267, 16.476
>   50 => 17
>   Next segment small: 51, next 0.016 < 1.000
>   Replaced 17 with: 52, at 47.440, 8.365
>   53 => 18
>   Next segment small: 54, next 0.016 < 1.000
>   Replaced 18 with: 55, at 48.172, 0.000
>   56 => 19
>   Next segment small: 57, next 0.016 < 1.000
>   Replaced 19 with: 58, at 47.440, -8.365
>   59 => 20
>   Next segment small: 60, next 0.016 < 1.000
>   Replaced 20 with: 61, at 45.267, -16.476
>   62 => 21
>   Next segment small: 63, next 0.016 < 1.000
>   Replaced 21 with: 64, at 41.718, -24.086
>   65 => 22
>   Next segment small: 66, next 0.016 < 1.000
>   Replaced 22 with: 67, at 36.902, -30.964
>   68 => 23
>   Next segment small: 69, next 0.016 < 1.000
>   Replaced 23 with: 70, at 30.964, -36.902
>   71 => 24
>   Next segment small: 72, next 0.016 < 1.000
>   Replaced 24 with: 73, at 24.086, -41.718
>   74 => 25
>   Next segment small: 75, next 0.016 < 1.000
>   Replaced 25 with: 76, at 16.476, -45.267
>   77 => 26
>   Next segment small: 78, next 0.016 < 1.000
>   Replaced 26 with: 79, at 8.365, -47.440
>   80 => 27
>   Next segment small: 81, next 0.016 < 1.000
>   Replaced 27 with: 82, at 0.000, -48.172
>   83 => 28
>   Next segment small: 84, next 0.016 < 1.000
>   Replaced 28 with: 85, at -8.365, -47.440
>   86 => 29
>   Next segment small: 87, next 0.016 < 1.000
>   Replaced 29 with: 88, at -16.476, -45.267
>   89 => 30
>   Next segment small: 90, next 0.016 < 1.000
>   Replaced 30 with: 91, at -24.086, -41.718
>   92 => 31
>   Next segment small: 93, next 0.016 < 1.000
>   Replaced 31 with: 94, at -30.964, -36.902
>   95 => 32
>   Next segment small: 96, next 0.016 < 1.000
>   Replaced 32 with: 97, at -36.902, -30.964
>   98 => 33
>   Next segment small: 99, next 0.016 < 1.000
>   Replaced 33 with: 100, at -41.718, -24.086
>   101 => 34
>   Next segment small: 102, next 0.016 < 1.000
>   Replaced 34 with: 103, at -45.267, -16.476
>   104 => 35
>   Next segment small: 105, next 0.016 < 1.000
>   Replaced 35 with: 106, at -47.440, -8.365
>   107 => 36
>   Replaced 36 with: 108, at -48.172, -0.000
> AFTER ExtrusionLine::simplify on 37 points:
>   0: -48.172, -0.000, w 0.368, len 0.000
>   1: -47.440, 8.365, w 0.368, len 8.397
>   2: -45.267, 16.476, w 0.368, len 8.397
>   3: -41.718, 24.086, w 0.368, len 8.397
>   4: -36.902, 30.964, w 0.368, len 8.397
>   5: -30.964, 36.902, w 0.368, len 8.397
>   6: -24.086, 41.718, w 0.368, len 8.397
>   7: -16.476, 45.267, w 0.368, len 8.397
>   8: -8.365, 47.440, w 0.368, len 8.397
>   9: 0.000, 48.172, w 0.368, len 8.397
>   10: 8.365, 47.440, w 0.368, len 8.397
>   11: 16.476, 45.267, w 0.368, len 8.397
>   12: 24.086, 41.718, w 0.368, len 8.397
>   13: 30.964, 36.902, w 0.368, len 8.397
>   14: 36.902, 30.964, w 0.368, len 8.397
>   15: 41.718, 24.086, w 0.368, len 8.397
>   16: 45.267, 16.476, w 0.368, len 8.397
>   17: 47.440, 8.365, w 0.368, len 8.397
>   18: 48.172, 0.000, w 0.368, len 8.397
>   19: 47.440, -8.365, w 0.368, len 8.397
>   20: 45.267, -16.476, w 0.368, len 8.397
>   21: 41.718, -24.086, w 0.368, len 8.397
>   22: 36.902, -30.964, w 0.368, len 8.397
>   23: 30.964, -36.902, w 0.368, len 8.397
>   24: 24.086, -41.718, w 0.368, len 8.397
>   25: 16.476, -45.267, w 0.368, len 8.397
>   26: 8.365, -47.440, w 0.368, len 8.397
>   27: 0.000, -48.172, w 0.368, len 8.397
>   28: -8.365, -47.440, w 0.368, len 8.397
>   29: -16.476, -45.267, w 0.368, len 8.397
>   30: -24.086, -41.718, w 0.368, len 8.397
>   31: -30.964, -36.902, w 0.368, len 8.397
>   32: -36.902, -30.964, w 0.368, len 8.397
>   33: -41.718, -24.086, w 0.368, len 8.397
>   34: -45.267, -16.476, w 0.368, len 8.397
>   35: -47.440, -8.365, w 0.368, len 8.397
>   36: -48.172, -0.000, w 0.368, len 8.397
> ```
> 
> For each point 'len' is the distance to the previous point. The first point just says 'len 0' since there is no previous point. In any case, I see no tiny segment on this perimeter, so I suspect it comes from some later processing stage. It probably does make the seam slightly worse on that layer.

I agree.
Looks like this is a unrelated issue

SoftFever reviewed (approved):
Looks good.
Thank you very much

SoftFever commented:
Hi @sethml 
Just a quick heads up. 
I have temporarily reverted this change due to multiple reported crashes caused by the changes. 
A brief investigation indicates that the issue arose because, after simplification, a loop was degraded to a single line.
